File: ui/crypto.js
// ─── Browser-side AES-GCM helpers (Web Crypto) ─────────────────────────────
//
// The data key is derived from the user's password + the backend-supplied
// salt via PBKDF2-SHA256 (matching backend/crypto.py exactly). Both sides
// derive identical keys without ever transmitting the key itself.
//
// The CryptoKey is created `extractable: false`, so even an XSS-injected
// script can't read its raw bytes — only call `subtle.encrypt/decrypt`
// while the JS context is alive. On logout we drop the IDB record and the
// key reference; the browser GC's the underlying material soon after.
//
// File format on the wire is identical to backend/crypto.py:
//     [12-byte IV][ciphertext + 16-byte GCM tag]
// ─────────────────────────────────────────────────────────────────────────

export const PBKDF2_DEFAULT_ITERATIONS = 200_000;
const NONCE_LEN = 12;
const CANARY_PLAINTEXT = new TextEncoder().encode("FORGE-CANARY-V1");

const _utf8 = new TextEncoder();

// Convert a hex string (the form the backend serialises bytes in) into a
// Uint8Array.
export function hexToBytes(hex) {
  if (!hex) return new Uint8Array();
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) out[i] = parseInt(hex.substr(i * 2, 2), 16);
  return out;
}

export async function deriveKey(password, salt, iterations = PBKDF2_DEFAULT_ITERATIONS) {
  // Use a Uint8Array we can zero after PBKDF2. The `password` string
  // parameter is still in the JS heap somewhere outside our reach — that's
  // a Python/JS limitation, not a bug. This still beats keeping a buffer
  // alive across the rest of the function.
  const pwBytes = _utf8.encode(password);
  let baseKey;
  try {
    baseKey = await crypto.subtle.importKey(
      "raw", pwBytes, { name: "PBKDF2" }, /* extractable */ false, ["deriveKey"]
    );
  } finally {
    pwBytes.fill(0);   // best-effort wipe of the raw password buffer
  }
  return crypto.subtle.deriveKey(
    {
      name: "PBKDF2",
      salt: salt instanceof Uint8Array ? salt : new Uint8Array(salt),
      iterations,
      hash: "SHA-256",
    },
    baseKey,
    { name: "AES-GCM", length: 256 },
    /* extractable */ false,
    ["encrypt", "decrypt"]
  );
}

// Encrypt arbitrary bytes (Uint8Array or ArrayBuffer). Returns Uint8Array
// matching the backend's "[iv][ct+tag]" format.
export async function encryptBytes(key, plaintext) {
  const iv = crypto.getRandomValues(new Uint8Array(NONCE_LEN));
  const ct = await crypto.subtle.encrypt({ name: "AES-GCM", iv }, key, plaintext);
  const out = new Uint8Array(NONCE_LEN + ct.byteLength);
  out.set(iv, 0);
  out.set(new Uint8Array(ct), NONCE_LEN);
  return out;
}

export async function decryptBytes(key, blob) {
  const data = blob instanceof Uint8Array ? blob : new Uint8Array(blob);
  if (data.byteLength < NONCE_LEN + 16) throw new Error("ciphertext too short");
  const iv = data.subarray(0, NONCE_LEN);
  const ct = data.subarray(NONCE_LEN);
  return await crypto.subtle.decrypt({ name: "AES-GCM", iv }, key, ct);
}

// Encrypt a Blob (file picker output) and return a new Blob containing the
// canonical "[iv][ct+tag]" bytes. Used by the upload path so plaintext
// never crosses the wire.
export async function encryptBlob(key, blob) {
  const buf = await blob.arrayBuffer();
  const out = await encryptBytes(key, buf);
  return new Blob([out], { type: "application/octet-stream" });
}

// Verify that a freshly-derived key matches the stored canary. Returns
// true if decryption succeeds AND yields the canonical canary plaintext.
export async function verifyCanary(key, canaryHex) {
  if (!canaryHex) return true;   // no canary stored yet (fresh setup)
  try {
    const pt = new Uint8Array(await decryptBytes(key, hexToBytes(canaryHex)));
    if (pt.length !== CANARY_PLAINTEXT.length) return false;
    for (let i = 0; i < pt.length; i++) {
      if (pt[i] !== CANARY_PLAINTEXT[i]) return false;
    }
    return true;
  } catch {
    return false;
  }
}

File: backend/crypto.py
"""AES-256-GCM helpers for at-rest asset encryption.

Used by both the FastAPI backend and runner subprocesses so refs and outputs
never touch disk in plaintext. The 32-byte data key is derived from the
user's password via PBKDF2 and only ever lives in RAM:

  - Backend keeps it on the `_Auth` singleton after setup/login/unlock.
  - Runners receive it as `FORGE_DATA_KEY` (hex) on spawn.

On-disk format for an encrypted file:
    [12-byte nonce] [ciphertext + 16-byte GCM tag]

There's no version header — `.enc` files are assumed to be this format.
If you ever change the format, bump the file extension.
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Optional

# `cryptography` is a hard dep when encryption is enabled. Imported lazily
# so the module still loads on hosts that haven't installed it yet (the
# legacy unencrypted path will still work for asset listing, etc.).
def _aesgcm():
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    return AESGCM


# ── Key derivation ────────────────────────────────────────────────────────
PBKDF2_ITERATIONS = 200_000  # ~200ms on a modern CPU; tune up if too fast
SALT_LEN          = 16
NONCE_LEN         = 12
CANARY_PLAINTEXT  = b"FORGE-CANARY-V1"   # known plaintext, encrypted at setup


def derive_key(password: str, salt: bytes) -> bytes:
    """Derive a 32-byte AES-256 key from a password + salt via PBKDF2-SHA256.

    Best-effort: we copy the password into a mutable bytearray and zero
    that buffer after derivation. Doesn't reach the original `password`
    string (Python's str immutability + interning) but reduces the number
    of plaintext copies hanging around in our process.
    """
    if len(salt) < 8:
        raise ValueError("salt too short")
    import hashlib
    pw = bytearray(password.encode("utf-8"))
    try:
        return hashlib.pbkdf2_hmac("sha256", pw, salt, PBKDF2_ITERATIONS, dklen=32)
    finally:
        for i in range(len(pw)):
            pw[i] = 0


def new_salt() -> bytes:
    return os.urandom(SALT_LEN)


# ── Bytes-level encrypt / decrypt ─────────────────────────────────────────
def encrypt_bytes(key: bytes, plaintext: bytes) -> bytes:
    """Encrypt with AES-GCM. Returns nonce || ciphertext || tag."""
    nonce = os.urandom(NONCE_LEN)
    ct_with_tag = _aesgcm()(key).encrypt(nonce, plaintext, None)
    return nonce + ct_with_tag


def decrypt_bytes(key: bytes, blob: bytes) -> bytes:
    """Decrypt the format produced by encrypt_bytes(). Raises on bad key/tampering."""
    if len(blob) < NONCE_LEN + 16:
        raise ValueError("ciphertext too short")
    nonce, ct_with_tag = blob[:NONCE_LEN], blob[NONCE_LEN:]
    return _aesgcm()(key).decrypt(nonce, ct_with_tag, None)


# ── Canary helpers (verifies the password without exposing the key) ──────
def make_canary(key: bytes) -> bytes:
    """Encrypt a known plaintext so a future unlock attempt can verify the key."""
    return encrypt_bytes(key, CANARY_PLAINTEXT)


def verify_canary(key: bytes, blob: bytes) -> bool:
    try:
        return decrypt_bytes(key, blob) == CANARY_PLAINTEXT
    except Exception:
        return False


# ── File-level encrypt / decrypt ──────────────────────────────────────────
def encrypted_path(p: Path) -> Path:
    """Append .enc unless already present."""
    return p if p.suffix == ".enc" else p.with_suffix(p.suffix + ".enc")


def visible_path(p: Path) -> Path:
    """Strip the trailing .enc, if any, for display/UI use."""
    return p.with_suffix("") if p.suffix == ".enc" else p


def find_on_disk(visible: Path) -> Optional[Path]:
    """Given the user-facing path, return whichever of <name>.enc or <name> exists."""
    enc = encrypted_path(visible)
    if enc.exists():
        return enc
    if visible.exists():
        return visible
    return None


def write_encrypted(key: bytes, dest_visible: Path, plaintext: bytes) -> Path:
    """Encrypt and write to <dest>.enc atomically. Returns the on-disk path."""
    dest = encrypted_path(dest_visible)
    dest.parent.mkdir(parents=True, exist_ok=True)
    blob = encrypt_bytes(key, plaintext)
    tmp = dest.with_suffix(dest.suffix + ".tmp")
    with open(tmp, "wb") as f:
        f.write(blob)
    os.replace(tmp, dest)
    return dest


def read_decrypted(key: bytes, source_visible: Path) -> bytes:
    """Read whichever of <name>.enc or <name> exists; decrypt the encrypted form."""
    on_disk = find_on_disk(source_visible)
    if on_disk is None:
        raise FileNotFoundError(source_visible)
    raw = on_disk.read_bytes()
    if on_disk.suffix == ".enc":
        return decrypt_bytes(key, raw)
    return raw   # legacy plaintext (kept for migration / dev convenience)


def looks_encrypted(path: Path) -> bool:
    """True if the file is at <name>.enc on disk (not a content check — by name)."""
    return path.suffix == ".enc"
